File: back/src/main/java/com/funstat/store/MdDao.java
package com.funstat.store;

import com.funstat.domain.Ohlc;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.transaction.support.TransactionTemplate;
import org.sqlite.SQLiteDataSource;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class MdDao {

    private final DataSourceTransactionManager manager;
    final SQLiteDataSource ds;

    ConcurrentHashMap<String, Boolean> tableCreated = new ConcurrentHashMap<>();

    public MdDao(SQLiteDataSource ds) {
        this.ds = ds;
        this.manager = new DataSourceTransactionManager(ds);
    }


    private void saveInTransaction(String sql, List<Map<String, Object>> data) {
        NamedParameterJdbcTemplate namedTemplate = new NamedParameterJdbcTemplate(ds);
        new TransactionTemplate(manager).execute(status -> {
            long start = System.currentTimeMillis();
            namedTemplate.batchUpdate(sql, data.toArray(new Map[0]));
            double dur = (System.currentTimeMillis() - start) / 1000.0;
            System.out.println("inserting " + data.size() + " took " + dur + " sec ," + " rate is " +
                    data.size() / dur + " per sec");
            return null;
        });
    }


    public void insert(List<firelib.domain.Ohlc> ohlcs, String table) {
        ensureExist(table);
        List<Map<String, Object>> data = ohlcs.stream().map(oh -> {
            return new HashMap<String, Object>() {
                {
                    put("DT", Timestamp.valueOf(LocalDateTime.ofInstant(oh.getDtGmtEnd(), ZoneOffset.UTC)));
                    put("OPEN", oh.getO());
                    put("HIGH", oh.getH());
                    put("LOW", oh.getL());
                    put("CLOSE", oh.getC());
                }
            };
        }).collect(Collectors.toList());

        try {
            saveInTransaction("insert into " + table + "(DT,O,H,L,C) values (:DT,:OPEN,:HIGH,:LOW,:CLOSE)", data);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void insertJ(List<Ohlc> ohlcs, String table) {
        List<firelib.domain.Ohlc> conv = ohlcs.stream().map(oh -> {
            return new firelib.domain.Ohlc(oh.close, oh.dateTime.toInstant(ZoneOffset.UTC), oh.high, oh.low, oh.open, 0,0);
        }).collect(Collectors.toList());
        insert(conv,table);
    }

    void ensureExist(String table) {
        tableCreated.computeIfAbsent(table, (k) -> {
            JdbcTemplate template = new JdbcTemplate(ds);
            template
                    .execute("create table if not exists " + table +
                            " (dt TIMESTAMPTZ, " +
                            "o DOUBLE PRECISION  not NULL," +
                            "h DOUBLE PRECISION  not NULL," +
                            "l DOUBLE PRECISION  not NULL," +
                            "c DOUBLE PRECISION  not NULL," +
                            "primary key (dt)) ;");

            return true;
        });
    }

    private Ohlc mapOhlc(ResultSet rs) throws SQLException {
        return new Ohlc(rs.getTimestamp("DT").toLocalDateTime(), rs.getDouble("o"), rs.getDouble("h"), rs.getDouble("l"), rs.getDouble("c"));
    }

    public Optional<Ohlc> queryLast(String code) {
        List<Ohlc> ret = new NamedParameterJdbcTemplate(ds).query("select * from " + code + " order by dt desc LIMIT 1 ", (rs, rowNum) -> mapOhlc(rs));
        return ret.size() == 0 ? Optional.empty() : Optional.of(ret.get(0));
    }

    public List<Ohlc> queryAll(String code) {
        return new NamedParameterJdbcTemplate(ds).query("select * from " + code + " order by dt asc ", (rs, rowNum) -> mapOhlc(rs));
    }

}

File: back/src/main/java/com/funstat/store/MdDao.kt
package com.funstat.store

import com.funstat.finam.FinamDownloader
import firelib.common.interval.Interval
import firelib.common.misc.atUtc
import firelib.common.misc.toInstantDefault
import firelib.common.reader.SimplifiedReader
import firelib.domain.Ohlc
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.transaction.support.TransactionTemplate
import org.sqlite.SQLiteDataSource
import java.sql.ResultSet
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class MdDao(internal val ds: SQLiteDataSource) {

    private val manager: DataSourceTransactionManager = DataSourceTransactionManager(ds)

    internal var tableCreated = ConcurrentHashMap<String, Boolean>()

    private fun saveInTransaction(sql: String, data: List<Map<String, Any>>) {

        TransactionTemplate(manager).execute<Any> {
            val start = System.currentTimeMillis()
            NamedParameterJdbcTemplate(ds).batchUpdate(sql, data.toTypedArray())
            val dur = (System.currentTimeMillis() - start) / 1000.0
            println("MdDao: inserting " + data.size + " took " + dur + " sec ," + " rate is " +
                    data.size / dur + " per sec")
            null
        }
    }


    fun insertOhlc(ohlcs: List<Ohlc>, tableIn: String) {
        val table = normName(tableIn)
        ensureExist(table)
        val data = ohlcs.filter {
            if (it.open.isFinite()) {
                true
            } else {
                println("not correct ${it}")
                false
            }
        }.map {
            mapOf("DT" to it.endTime.toEpochMilli(),
                    "OPEN" to it.open,
                    "HIGH" to it.high,
                    "LOW" to it.low,
                    "CLOSE" to it.close,
                    "VOLUME" to it.volume
            )
        }

        try {
            saveInTransaction("insert or replace into $table(DT,O,H,L,C,V) values (:DT,:OPEN,:HIGH,:LOW,:CLOSE,:VOLUME)", data)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    internal fun ensureExist(table: String) {
        tableCreated.computeIfAbsent(table) { k ->
            JdbcTemplate(ds).execute("create table if not exists ${table}"  +
                            " (dt INTEGER not NULL, " +
                            "o DOUBLE PRECISION  not NULL," +
                            "h DOUBLE PRECISION  not NULL," +
                            "l DOUBLE PRECISION  not NULL," +
                            "c DOUBLE PRECISION  not NULL," +
                            "v INT not NULL," +
                            "primary key (dt)) ;")

            true
        }
    }

    private fun mapOhlc(rs: ResultSet): Ohlc {
        return Ohlc(rs.getTimestamp("DT").toInstant(), rs.getDouble("o"), rs.getDouble("h"), rs.getDouble("l"), rs.getDouble("c"), volume = rs.getLong("v"), interpolated = false)
    }

    fun normName(name: String): String {
        return name.replace('-', '_');
    }

    fun queryLast(codeIn: String): Optional<Ohlc> {
        val code = normName(codeIn)
        ensureExist(code)
        val ret = NamedParameterJdbcTemplate(ds).query("select * from $code order by dt desc LIMIT 1 ") { rs, rowNum -> mapOhlc(rs) }
        return if (ret.size == 0) Optional.empty() else Optional.of(ret[0])
    }

    fun queryAll(codeIn: String): List<Ohlc> {
        return queryAll(codeIn, LocalDateTime.ofEpochSecond(0, 0, ZoneOffset.UTC))
    }

    fun queryAll(codeIn: String, start: LocalDateTime, limit : Int = 10_000_000): List<Ohlc> {
        val code = normName(codeIn)
        ensureExist(code)

        val map = mapOf("DT" to  start.toInstantDefault().toEpochMilli(),
                "LIMIT" to limit
        )
        return NamedParameterJdbcTemplate(ds).query("select * from $code where dt > :DT order by dt asc LIMIT :LIMIT", map, { rs, _ -> mapOhlc(rs) })
    }


}

class SimplifiedReaderImpl(val mdDao: MdDao, val code : String, val startTime : Instant) : SimplifiedReader{

    var lastRead : Instant = startTime

    var buffer  = LinkedList<Ohlc>()

    fun read(){
        println("doing thing")
        val list = mdDao.queryAll(code, lastRead.atUtc(), 20_000)
        if(list.isNotEmpty()){
            buffer.addAll(list)
            lastRead = list.last().endTime
        }
    }

    override fun peek(): Ohlc? {
        if(buffer.isEmpty()){
            read()
        }
        return buffer.peek()
    }

    override fun poll(): Ohlc {
        return buffer.poll()
    }

}



fun main(){
    val impl = MdStorageImpl()
    val dao = impl.getDao(FinamDownloader.SOURCE, Interval.Min10.name)

    val start = LocalDateTime.now().minusDays(200).toInstantDefault()
    val reader = SimplifiedReaderImpl(dao, "sber", start)
    while(reader.peek() != null){
        println("ohlc ${reader.poll()}")
    }
}